src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,199 @@
+        }
+
+        public static void RegisterFactory(string providerInvariantName, string factoryTypeAssemblyQualifiedName)
+        {
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            ADP.CheckArgumentLength(factoryTypeAssemblyQualifiedName, nameof(factoryTypeAssemblyQualifiedName));
+            
+            // this method performs a deferred registration: the type name specified is checked when the factory is requested for the first time. 
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factoryTypeAssemblyQualifiedName, null);
+        }
+        
+        public static void RegisterFactory(string providerInvariantName, Type providerFactoryClass)
+        {
+            RegisterFactory(providerInvariantName, GetFactoryInstance(providerFactoryClass));
+        }
+
+        public static void RegisterFactory(string providerInvariantName, DbProviderFactory factory)
+        {
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            ADP.CheckArgumentNull(factory, nameof(factory));
+
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factory.GetType().AssemblyQualifiedName, factory);
+        }
+        
+        public static bool UnregisterFactory(string providerInvariantName)
+        {
+            return !string.IsNullOrWhiteSpace(providerInvariantName) && _registeredFactories.TryRemove(providerInvariantName, out ProviderRegistration providerRegistration);
+        }
+        
+        private static DbProviderFactory GetFactory(string providerInvariantName, bool throwOnError)
+        {
+            if (throwOnError)
+            {
+                ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            }
+            else
+            {
+                if (string.IsNullOrWhiteSpace(providerInvariantName))
+                {
+                    return null;
+                }
+            }
+            bool wasRegistered = _registeredFactories.TryGetValue(providerInvariantName, out ProviderRegistration registration);
+            if (!wasRegistered)
+            {
+                return throwOnError ? throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_InvariantNameNotFound, providerInvariantName)) : (DbProviderFactory)null;
+            }
+            DbProviderFactory toReturn = registration.FactoryInstance;
+            if (toReturn == null)
+            {
+                // Deferred registration, do checks now on the type specified and register instance in storage.
+                // Even in the case of throwOnError being false, this will throw when an exception occurs checking the registered type as the user has to be notified the 
+                // registration is invalid, even though the registration is there.
+                Type providerType = Type.GetType(registration.FactoryTypeAssemblyQualifiedName);
```

Comment:
Do we need a check for null or empty for `registration.FactoryTypeAssemblyQualifiedName` ? 
Or is it that `Type.GetType` would return a null for an null or empty parameter passed in?

Reply:
The check for that is done at RegisterFactory(string, string), line 114: 
```cs
        public static void RegisterFactory(string providerInvariantName, string factoryTypeAssemblyQualifiedName)
        {
            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
            ADP.CheckArgumentLength(factoryTypeAssemblyQualifiedName, nameof(factoryTypeAssemblyQualifiedName));
```

It then creates a ProviderRegistration class which on its own again checks for the value being not null and have a length, so it's not possible to have a name that's null or empty.

Reply:
Not a big deal, but my first thought when I saw the ProviderRegistration constructor is that it could validate that either the instance of the AQN are not null. If the instance is passed, the only case in which we may need to obtain the AQN again is GetProviderClasses, which should be used infrequently.

FWIW, even if we did that, this code would still be safe without a null check.

Reply:
the struct is private and its ctor is used only in codepaths which indeed check inputs, the null check in the ProviderRegistration could currently be removed. I added it there for clarity as it shouldn't be null. It's easier for maintainers who might alter the code later on: they then don't have to worry if a code path could lead to a null value there. btw the netfx code contains duplicate null checks too in a code path, which I guess are there for similar reasons :)

Reply:
I get all that, and I like that you put the check there... maybe I am not getting my main point across: if the instance is set, we can allow the AQN to be null.

Reply:
ah I didn't understand that from your remark. If the instance is null, then indeed the name can be null, as the if at line 158 isn't true. But what shall I do now: wrap the check in an if (complexer code) ? Or leave it as is? Or remove the null check?

Reply:
I don't feel strongly about this but thinking about it again, it seems that doing the validation on the constructor provides little value. E.g. there is still a default constructor that allows to create an invalid ProviderRegistration, and the validation hast to be performed elsewhere anyway. 

I still believe we don't need to eagerly store the assembly qualified name in the ProviderRegistration when the instance is provided (we can compute it on the fly on GetFactoryClasses), but I don't feel super strongly about that either.

Reply:
The default ctor is indeed a mitigating factor for the check in the other ctor. I'll remove that one, and leave it at that. storing the aqn isn't harmful in any way, in fact, explicitly storing a null might raise an eyebrow as the question might arise "why is it deliberately left empty" which then requires a comment "we can recompute it later", IMHO unneeded complexity.